Storing loginStatus value in a database

pooneh-nb opened the pull request:
#### 96606df018fb75b2e6f5b643438712c2f642eafb
<pre>
Storing loginStatus value in a database
<a href="https://bugs.webkit.org/show_bug.cgi?id=279162">https://bugs.webkit.org/show_bug.cgi?id=279162</a>
<a href="https://rdar.apple.com/135310202">rdar://135310202</a>

Reviewed by Chris Dumez and Alex Christensen.

In the initial implementation, the loginStatus was stored in a temporary hashmap, and the status was cleared once the session ended. In this patch, we update the implementation to store the status persistently in the LoginStatus table within the observations.db database.
This functionality is covered by the layout test http/tests/login-status/login-status-api-getter-setter-functions.html.

* Source/WebCore/page/LoginStatus.cpp:
(WebCore::LoginStatus::isolatedCopy const):
(WebCore::LoginStatus::isolatedCopy):
* Source/WebCore/page/LoginStatus.h:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::expectedTableAndIndexQueries):
(WebKit::ResourceLoadStatisticsStore::sortedTables):
(WebKit::ResourceLoadStatisticsStore::createSchema):
(WebKit::ResourceLoadStatisticsStore::destroyStatements):
(WebKit::ResourceLoadStatisticsStore::insertLoginStatus):
(WebKit::ResourceLoadStatisticsStore::removeLoginStatus):
(WebKit::ResourceLoadStatisticsStore::loginStatusExists):
(WebKit::ResourceLoadStatisticsStore::domainID const):
(WebKit::ResourceLoadStatisticsStore::loginStatusDomainID const):
(WebKit::ResourceLoadStatisticsStore::domainIDWithTable const):
(WebKit::ResourceLoadStatisticsStore::setLoginStatus):
(WebKit::ResourceLoadStatisticsStore::isLoggedIn):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::setLoginStatus):
(WebKit::WebResourceLoadStatisticsStore::isLoggedIn):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:

Canonical link: <a href="https://commits.webkit.org/283505@main">https://commits.webkit.org/283505@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc39479d1c8e12ab7f67cc5a5188ac302bdd28a2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66449 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45824 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19070 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70482 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17582 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68567 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53623 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17342 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53295 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11892 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69516 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42232 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57521 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33949 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38903 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14909 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15935 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60806 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15251 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72185 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10406 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14624 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60629 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10438 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57590 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60943 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8589 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2201 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10076 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41631 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42708 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43891 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42451 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -229,6 +235,19 @@ constexpr auto createSubresourceUniqueRedirectsFrom = "CREATE TABLE SubresourceU
 constexpr auto createOperatingDates = "CREATE TABLE OperatingDates ("
     "year INTEGER NOT NULL, month INTEGER NOT NULL, monthDay INTEGER NOT NULL)"_s;
 
+    // constexpr auto createObservedDomain = "CREATE TABLE ObservedDomains ("
```
Remove comment.

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1216,7 +1245,8 @@ bool ResourceLoadStatisticsStore::insertObservedDomain(const ResourceLoadStatist
         || scopedStatement->bindInt(IsScheduledForAllButCookieDataRemovalIndex, loadStatistics.gotLinkDecorationFromPrevalentResource
-        || scopedStatement->bindDouble(MostRecentWebPushInteractionTimeIndex, 0.0)) != SQLITE_OK) {
+        || scopedStatement->bindDouble(MostRecentWebPushInteractionTimeIndex, 0.0)) != SQLITE_OK
+        || scopedStatement->step() != SQLITE_DONE) {
```
This is redundant with line 1254.  Let's not make this change.

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1228,6 +1258,70 @@ bool ResourceLoadStatisticsStore::insertObservedDomain(const ResourceLoadStatist
+
+    auto scopedStatement = this->scopedStatement(m_insertLoginstatusStatement, insertLoginStatusQuery, "insertLoginstatus"_s);
+    if (!scopedStatement
+        || scopedStatement->bindText(1, domain.string()) != SQLITE_OK
```
Could we give 1..7 names like ResourceLoadStatisticsStore::insertObservedDomain has?

pooneh-nb replied:
I created an enum in function scope for the ResourceLoadStatisticsStore::insertLoginStatus

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1228,6 +1258,70 @@ bool ResourceLoadStatisticsStore::insertObservedDomain(const ResourceLoadStatist
+{
+    ASSERT(!RunLoop::isMain());
+
+    auto domainIDToRemove = loginStatusdomainID(WTFMove(domain));
```
D should probably be capitalized.

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1753,6 +1861,31 @@ void ResourceLoadStatisticsStore::requestStorageAccess(SubFrameDomain&& subFrame
+
+void ResourceLoadStatisticsStore::isLoggedIn(RegistrableDomain&& domain, CompletionHandler<void(bool)>&& completionHandler)
+{
+    ASSERT(RunLoop::isMain());
```
This is incorrect.

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1753,6 +1861,31 @@ void ResourceLoadStatisticsStore::requestStorageAccess(SubFrameDomain&& subFrame
+{
+    ASSERT(RunLoop::isMain());
+
+    completionHandler(loginStatusExists(WTFMove(domain)));
```
Since this doesn't schedule anything on other threads, I think we can just return the value instead of using a CompletionHandler.

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -449,23 +448,22 @@ void WebResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain,
-    auto it = m_loginStatus.find(domain);
-    completionHandler(it != m_loginStatus.end() && it->value.first == IsLoggedIn::LoggedIn);
+    if (m_statisticsStore)
+        m_statisticsStore->isLoggedIn(WTFMove(domain), WTFMove(completionHandler));
```
This needs a call to postTask to call isLoggedIn on a background thread then postTaskReply to call the completion handler.  That way everything will be on the right thread.

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1753,6 +1861,31 @@ void ResourceLoadStatisticsStore::requestStorageAccess(SubFrameDomain&& subFrame
+
+    auto loginStatusToSet = lastAuthentication && lastAuthentication->hasExpired() ? std::nullopt : std::optional(WTFMove(lastAuthentication));
+    if (loginStatusToSet)
+        loginStatusToSet->setTimeToLive(WebCore::LoginStatus::TimeToLiveLong);
```
It would be nice to eventually have a test for this.

achristensen07 reviewed (changes requested):
Could you mention in the change log that this functionality is covered by the layout test http/tests/login-status/login-status-api-getter-setter-functions.html ?
Also, run that test in a debug build to make sure no assertions are hit.

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -449,23 +448,22 @@ void WebResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain,
     ASSERT(RunLoop::isMain());
 
-    auto it = m_loginStatus.find(domain);
-    completionHandler(it != m_loginStatus.end() && it->value.first == IsLoggedIn::LoggedIn);
```
We should remove `m_loginStatus` since it is no longer being used.

pooneh-nb replied:
These lines have been removed actually

charliewolfe replied:
I meant the variable that is still being stored on this object (`m_loginStatus` in the header).

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -449,23 +448,22 @@ void WebResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain,
+        loginStatusToSet->setTimeToLive(WebCore::LoginStatus::TimeToLiveLong);
+
+    if (m_statisticsStore)
+        m_statisticsStore->setLoginStatus(WTFMove(domain), loggedInStatus, WTFMove(lastAuthentication), WTFMove(completionHandler));
```
We need to return here, otherwise we will call the completion handler twice.

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1753,6 +1861,31 @@ void ResourceLoadStatisticsStore::requestStorageAccess(SubFrameDomain&& subFrame
+    if (loggedInStatus == IsLoggedIn::LoggedIn) {
+        auto insertStatus = insertLoginStatus(WTFMove(domain), loggedInStatus, WTFMove(lastAuthentication));
+        if (!insertStatus)
+            return;
```
completion handler needs to be called here, also we don’t need the `insertStatus` variable

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1228,6 +1249,104 @@ bool ResourceLoadStatisticsStore::insertObservedDomain(const ResourceLoadStatist
+    return;
+}
+
+void ResourceLoadStatisticsStore::removeLoginStatus(const RegistrableDomain&& domain)
```
This needs to be non const to fix the Linux build.

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
```diff
@@ -409,6 +419,11 @@ class ResourceLoadStatisticsStore final : public DatabaseUtilities, public CanMa
+    std::unique_ptr<WebCore::SQLiteStatement> m_insertLoginstatusStatement;
+    mutable std::unique_ptr<WebCore::SQLiteStatement> m_removeLoginStatusStatement;
+    mutable std::unique_ptr<WebCore::SQLiteStatement> m_loginstatusExistsStatement;
+
```
nit: extra newline.

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
```diff
@@ -52,7 +52,9 @@ template<> struct IsDeprecatedWeakRefSmartPointerException<WebKit::ResourceLoadS
 class KeyedEncoder;
+class LoginStatus;
 class SQLiteStatement;
+enum class IsLoggedIn : uint8_t;
```
In a separate PR we could make IsLoggedIn use bool as its underlying type.

achristensen07 commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1249,7 +1368,21 @@ std::optional<unsigned> ResourceLoadStatisticsStore::domainID(const RegistrableD
+{
+    ASSERT(!RunLoop::isMain());
+
+    return domainIDWithTable(domain, m_loginstatusDomainIDFromStringStatement, loginStatusdomainIDFromStringQuery);
```
Status should have a capital S, and Domain should have a capital D in both variable names.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1193,6 +1211,9 @@ void ResourceLoadStatisticsStore::destroyStatements()
     m_updateMostRecentWebPushInteractionTimeStatement = nullptr;
+    m_insertObservedDomainStatement = nullptr;
+    m_removeLoginStatusStatement = nullptr;
+    m_loginstatusExistsStatement = nullptr;
```
`m_loginstatus...` -> `m_loginStatus...` ?

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1228,6 +1249,104 @@ bool ResourceLoadStatisticsStore::insertObservedDomain(const ResourceLoadStatist
+    enum {
+        DomainIDIndex,
+        RegistrableDomainIndex,
+        UserNameIndex,
```
`UserNameIndex` -> `UsernameIndex`

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1228,6 +1249,104 @@ bool ResourceLoadStatisticsStore::insertObservedDomain(const ResourceLoadStatist
+    return;
+}
+
+void ResourceLoadStatisticsStore::removeLoginStatus(RegistrableDomain&& domain)
```
Seems all these functions do not need to take a `RegistrableDomain&&`, a `const RegistrableDomain&` would suffice. Also, `WTFMove(domain)` is pointless  to call `loginStatusDomainID()` since it takes in a `const RegistrableDomain&`.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -449,23 +448,38 @@ void WebResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain,
+    if (loginStatusToSet)
+        loginStatusToSet->setTimeToLive(WebCore::LoginStatus::TimeToLiveLong);
+
+    postTask([this, domain = WTFMove(domain), loggedInStatus, lastAuthentication = WTFMove(lastAuthentication), completionHandler = WTFMove(completionHandler)]() mutable {
```
That's a bug. postTask() dispatches to another thread so you need `crossThreadCopy()` the non-thread safe types you're capturing in the lambda:
```
domain = crossThreadCopy(WTFMove(domain)), lastAuthentication = crossTheadCopy(WTFMove(lastAuthentication))
```

Note that this will require adding an `isolatedCopy()` member function to `LoginStatus`.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -449,23 +448,38 @@ void WebResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain,
 
-    auto it = m_loginStatus.find(domain);
-    completionHandler(it != m_loginStatus.end() && it->value.first == IsLoggedIn::LoggedIn);
+    postTask([this, domain = WTFMove(domain), completionHandler = WTFMove(completionHandler)]() mutable {
```
BUG: `crossThreadCopy()` for the `domain`

cdumez reviewed (changes requested):
r- due to thread safety bugs.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1228,6 +1249,104 @@ bool ResourceLoadStatisticsStore::insertObservedDomain(const ResourceLoadStatist
+        ITP_RELEASE_LOG_DATABASE_ERROR("removeLoginstatus: failed to step statement");
+}
+
+bool ResourceLoadStatisticsStore::loginStatusExists(const RegistrableDomain&& domain)
```
Shouldn't be `const` and an r-value reference. Should just use `const RegistrableDomain&` as mentioned above. This is causing Linux build breakage.

cdumez commented on Source/WebCore/page/LoginStatus.h:
```diff
@@ -60,6 +60,9 @@ class LoginStatus {
     WallTime loggedInTime() const { return m_loggedInTime; }
     Seconds timeToLive() const { return m_timeToLive; }
 
+    LoginStatus isolatedCopy() const & { return LoginStatus { *this }; }
```
This doesn't look safe. It is not calling `isolatedCopy()` on any of the data members. m_domain & m_username definitely needs to be isolated copied since they are Strings (I didn't check the others in detail).

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1228,6 +1249,104 @@ bool ResourceLoadStatisticsStore::insertObservedDomain(const ResourceLoadStatist
+    auto scopedStatement = this->scopedStatement(m_removeLoginStatusStatement, removeLoginStatusQuery, "removeLoginStatus"_s);
+    if (!scopedStatement
+        || scopedStatement->bindInt(1, *domainIDToRemove) != SQLITE_OK
+        || scopedStatement->step() != SQLITE_DONE)
```
I think the coding style says we should use curly brackets if the condition spans several lines.

pooneh-nb replied:
Do you mean for the condition starts from line 1283? I'm using curly brackets here. I didn't get style error for this part. Am I missing something?

cdumez replied:
I mean on the line I commented on, there should be an opening curly bracket and there still isn't

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1228,6 +1249,104 @@ bool ResourceLoadStatisticsStore::insertObservedDomain(const ResourceLoadStatist
     return true;
 }
 
+void ResourceLoadStatisticsStore::insertLoginStatus(const RegistrableDomain& domain, IsLoggedIn loggedInStatus, std::optional<LoginStatus>&& lastAuthentication)
```
`const std::optional<LoginStatus>&` seems to suffice since you're not WTFMove() it internally.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1753,6 +1886,27 @@ void ResourceLoadStatisticsStore::requestStorageAccess(SubFrameDomain&& subFrame
+        loginStatusToSet->setTimeToLive(WebCore::LoginStatus::TimeToLiveLong);
+
+    if (loggedInStatus == IsLoggedIn::LoggedIn)
+        insertLoginStatus(WTFMove(domain), loggedInStatus, WTFMove(lastAuthentication));
```
WTFMove() is unnecessary for both parameters.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1753,6 +1886,27 @@ void ResourceLoadStatisticsStore::requestStorageAccess(SubFrameDomain&& subFrame
+    if (loggedInStatus == IsLoggedIn::LoggedIn)
+        insertLoginStatus(WTFMove(domain), loggedInStatus, WTFMove(lastAuthentication));
+    else if (loggedInStatus == IsLoggedIn::LoggedOut)
+        removeLoginStatus(WTFMove(domain));
```
WTFMove() is unnecessary.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1753,6 +1886,27 @@ void ResourceLoadStatisticsStore::requestStorageAccess(SubFrameDomain&& subFrame
+{
+    ASSERT(!RunLoop::isMain());
+
+    return loginStatusExists(WTFMove(domain));
```
`WTFMove()` is unnecessary, loginStatusExists() takes a `const RegistrableDomain&`

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1753,6 +1886,27 @@ void ResourceLoadStatisticsStore::requestStorageAccess(SubFrameDomain&& subFrame
+        removeLoginStatus(WTFMove(domain));
+}
+
+bool ResourceLoadStatisticsStore::isLoggedIn(RegistrableDomain&& domain)
```
Can take a `const RegistrableDomain&`. There is no perf benefit when using a r-value reference here it seems.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1753,6 +1886,27 @@ void ResourceLoadStatisticsStore::requestStorageAccess(SubFrameDomain&& subFrame
     });
 }
 
+void ResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain, IsLoggedIn loggedInStatus, std::optional<LoginStatus>&& lastAuthentication)
```
domain can be `const RegistrableDomain&`

cdumez commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -449,23 +448,38 @@ void WebResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain,
+            return;
+        }
+
+        m_statisticsStore->setLoginStatus(WTFMove(domain), loggedInStatus, WTFMove(lastAuthentication));
```
No need for `WTFMove()` for `domain`

cdumez commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -449,23 +448,38 @@ void WebResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain,
+        if (!m_statisticsStore) {
+            postTaskReply([completionHandler = WTFMove(completionHandler)]() mutable {
+                completionHandler(false);
+            });
```
Did you mean to early return after this? This looks like a bug.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -449,23 +448,39 @@ void WebResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain,
-        m_loginStatus.remove(domain);
-    completionHandler();
+    auto loginStatusToSet = lastAuthentication && lastAuthentication->hasExpired() ? std::nullopt : std::optional(WTFMove(lastAuthentication));
+    if (loginStatusToSet)
```
Here you check if `loginStatusToSet` is not nullopt, which seems fair given you may set it to nullopt on the previous line...

cdumez commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -449,23 +448,39 @@ void WebResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain,
+    if (loginStatusToSet)
+        loginStatusToSet->setTimeToLive(WebCore::LoginStatus::TimeToLiveLong);
+
+    postTask([this, domain = WTFMove(domain).isolatedCopy(), loggedInStatus, loginStatusToSet = WTFMove(*loginStatusToSet).isolatedCopy(), completionHandler = WTFMove(completionHandler)]() mutable {
```
... But here you dereference `loginStatusToSet` unconditionally, which would crash if `loginStatusToSet` was nullopt.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -449,23 +448,39 @@ void WebResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain,
+    if (loginStatusToSet)
+        loginStatusToSet->setTimeToLive(WebCore::LoginStatus::TimeToLiveLong);
+
+    postTask([this, domain = WTFMove(domain).isolatedCopy(), loggedInStatus, loginStatusToSet = WTFMove(loginStatusToSet)->isolatedCopy(), completionHandler = WTFMove(completionHandler)]() mutable {
```
This will still crash if `loginStatusToSet` is nullopt since you are dereferencing it. What you want is `crossThreadCopy(WTFMove(loginStatusToSet))`

Commit:
Storing loginStatus value in a database
https://bugs.webkit.org/show_bug.cgi?id=279162
rdar://135310202

Reviewed by Chris Dumez and Alex Christensen.

In the initial implementation, the loginStatus was stored in a temporary hashmap, and the status was cleared once the session ended. In this patch, we update the implementation to store the status persistently in the LoginStatus table within the observations.db database.
This functionality is covered by the layout test http/tests/login-status/login-status-api-getter-setter-functions.html.

* Source/WebCore/page/LoginStatus.cpp:
(WebCore::LoginStatus::isolatedCopy const):
(WebCore::LoginStatus::isolatedCopy):
* Source/WebCore/page/LoginStatus.h:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::expectedTableAndIndexQueries):
(WebKit::ResourceLoadStatisticsStore::sortedTables):
(WebKit::ResourceLoadStatisticsStore::createSchema):
(WebKit::ResourceLoadStatisticsStore::destroyStatements):
(WebKit::ResourceLoadStatisticsStore::insertLoginStatus):
(WebKit::ResourceLoadStatisticsStore::removeLoginStatus):
(WebKit::ResourceLoadStatisticsStore::loginStatusExists):
(WebKit::ResourceLoadStatisticsStore::domainID const):
(WebKit::ResourceLoadStatisticsStore::loginStatusDomainID const):
(WebKit::ResourceLoadStatisticsStore::domainIDWithTable const):
(WebKit::ResourceLoadStatisticsStore::setLoginStatus):
(WebKit::ResourceLoadStatisticsStore::isLoggedIn):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::setLoginStatus):
(WebKit::WebResourceLoadStatisticsStore::isLoggedIn):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:

Canonical link: https://commits.webkit.org/283505@main
